### PR TITLE
feat(rocm): V1 sampling ops for Hermes EngineCore

### PR DESCRIPTION
**Pull request by bakon11**
## Summary

Makes **Hermes Agent ↔ local Gemma-4-26B MoE FP8 on AMD ROCm (RDNA4)** finish chat turns *and* decode at usable speed when `temperature > 0`.

Public shout-out context: [@anothervariable](https://x.com/anothervariable/status/2086614087178293373) — Gemma 4 26B MoE on AMD ROCm via vllm.cpp.

### Root causes (lab)

**1) Missing ROCm V1 sampler ops (engine death)**  
After a large Hermes SOUL prefill (~38k tokens), sampling called V1 ops registered on CUDA/CPU only:

| Op | Symptom |
|----|---------|
| `kApplyAllowedTokenIds` (40) | title-gen allowlist → `no kernel for op 40 on device type 5` |
| `kApplyTemperature` (30) | temp/top_p path → `no kernel for op 30` |
| cascade | `request submitted to a stopped AsyncLLM` |

ROCm device type = 5. Missing registration = hard engine death, not a soft 4xx.

**2) `RandomSampleK` serial full-vocab scan (decode cliff)**  
HIP port initially matched CUDA's `<<<n, 1>>>` gumbel-max. On gfx1201 with Gemma-4 **V=262144**, every temp>0 token paid ~0.5s in sampling → Hermes decode **~1.75 t/s** vs greedy **~40–50 t/s**.

Fix: parallel block reduce, launch **`<<<n, kBlock>>>`** with `kBlock=256`.

### Changes

1. **`src/vt/rocm/rocm_sample.hip`** — HIP V1 sampler surface:
   - temperature, top-k/top-p, softmax / log-softmax
   - **parallel** gumbel-max random sample, penalties, min-p, logit bias
2. **`rocm_ops.hip` / `rocm_dense_basic.hip`** — register those + `ApplyAllowedTokenIds` / `ApplyTokenMask`
3. **Chat `max_tokens=-1`** (Hermes “unlimited”) → map to `VT_SERVER_MAX_NEW_TOKENS` default (**4096**) before `SamplingParams::PostInit` rejects `<1`
4. Docs: `ENVIRONMENT.md`, `FEATURES.md`, `USAGE.md` (sampler note)

No `vt::rocm` calls from `models/` (device-leakage clean). CUDA still has 1-thread `RandomSample` (hidden on fat SMs / smaller V); ROCm deliberately diverges for gfx1201.

### Lab verification (2× Radeon AI PRO R9700, gfx1201, ROCm 7.2)

| case | temp | decode t/s |
|------|------|----------:|
| short | 0.0 | ~51 |
| short | 0.7 | **~45** (was ~1.78) |
| hermesish+55 tools @~37k | 0.0 | ~42 |
| hermesish+55 tools @~37k | 0.7 | **~41** (was ~1.74) |

Also: title-gen allowlist + temp/top_p + `max_tokens=-1` → 200; Paris smoke OK.

Binary: CMake `OUTPUT_NAME` is `vllm-server` under `build-*/examples/`.

### Land order

1. **This PR (#234)** — ROCm V1 sampler + RandomSample perf
2. **#227** — KV fail-fast unschedulable waits
3. **#316** — SSE keepalives
4. **#317** — Gemma4/ROCm FP8 + SharedK-WMMA (split from closed #228)

### Follow-ups (not this PR)

- Optional CUDA RandomSample parity (same serial pattern exists in `cuda_sample.cu`)
- Prefill/decode stack beyond sampling (#317)

## Test plan

- [x] `python3 scripts/check-device-leakage.py`
- [x] `python3 scripts/check-env-doc.py`
- [x] `python3 scripts/check-doc-checkpoint.py --base origin/main --head HEAD`
- [x] `python3 scripts/check-pr-size.py` (under budget)
- [x] Live `:8010` Gemma-4-26B-A4B-it-fp8 dual-GPU: temp 0 vs 0.7 A/B
- [ ] CI green on tip-of-main

### CODE REVIEW

**Comment by bakon11** [comment]
## TESTERS WANTED — BF16 / FP16 **Gemma-4 MoE** on ROCm

Lab proof for this PR was **FP8** Gemma-4-26B-A4B MoE on **2× Radeon AI PRO R9700 (gfx1201)**. The code path here is **dtype-agnostic** (V1 sampler ops + Hermes `max_tokens=-1`), but we **cannot reliably soak BF16 MoE** on this box:

| Constraint | Lab |
|------------|-----|
| BF16 MoE weights | ~52 GiB |
| VRAM | 2×32 GiB — tight after KV/acts |
| Host RAM | ~30 GiB hard ceiling |
| Dense 12B BF16 | **not a substitute** (no expert routing / resident / prefill-batch MoE) |

### Please test if you have
- **Gemma-4-26B MoE** (or similar MoE) weights in **BF16 or FP16** (not only FP8)
- ROCm multi-GPU with enough VRAM (e.g. 2×48, 1×80+, or comfortable 2×32 with short ctx)
- Enough **host RAM** to load without thrashing

### Minimal matrix
1. Short chat: Paris / READY  
2. Hermes-like: `max_tokens=-1`, `temperature`/`top_p` > 0  
3. Title-gen style: `allowed_token_ids` present  
4. Medium stream (tools optional) — confirm **no** `no kernel for op … on device type 5` and no `stopped AsyncLLM` after first failure  

### Report back
- GPU(s) + ROCm version  
- Model id / dtype  
- `max_model_len` / `num_blocks`  
- Pass/fail per step + any log lines with `engine-fatal` / `no kernel for op`

Related: serve KV fail-fast **#227**; FP8 MoE prefill/resident package **#228** (FP8-proven; MoE structure notes there).

Thanks — BF16/FP16 MoE ROCm coverage has to come from the community; FP8 is what we can gate on RDNA4 consumer hardware day-to-day.

**Comment by bakon11** [comment]
## Addressing CHANGES_REQUESTED (localai-org-maint-bot)

### Automated sampler coverage
Added **CPU↔ROCm parity** cases in `tests/vt/test_ops_sample.cpp` (skip cleanly when ROCm is not linked):

| Case | Ops |
|------|-----|
| `ROCm greedy_argmax / temperature / top-k-p / allowed_ids / logprobs match CPU` | greedy, temperature, top-k, allowed_token_ids mask, logprobs |
| `ROCm apply_min_p / penalties surface matches CPU mask pattern` | min_p |

Lab run on this box (HIP build, dual R9700):

```text
./tests/test_ops_sample --test-case='*ROCm*'
test cases: 2 passed | assertions: 4108 passed | SUCCESS
```

CPU CI still exercises the existing hand-oracle sampler suite; ROCm block is opt-in like the CUDA parity block.

### Docs gates
- `docs/USAGE.md`: `max_tokens=-1` → `VT_SERVER_MAX_NEW_TOKENS`, ROCm V1 sampler surface note
- Local: `check-doc-checkpoint` OK, `check-agent-record` OK

### Merge order (unchanged)
Land **#234** first → rebase **#227** / **#228**.

Happy to add CUDA-style large-vocab ROCm greedy case or penalty/logit-bias matrix next if you want a wider surface before merge.

**Comment by bakon11** [comment]
Fixed agent-record: FEATURES.md Gemma4 ROCm cell was 258 chars (limit 220). Shortened; `check-public-doc-tables` OK.

**Comment by bakon11** [comment]
## Rebase + drop max_tokens clamp (addressing CHANGES_REQUESTED / main drift)

Rebuilt **tip-of-main** (`a0fa12c7`) as single commit `42ba78e2`.

### Dropped (now on main)
- `max_tokens=-1` → 4096 clamp in `protocol.cpp` / `serving_chat.cpp` — **main already** treats non-positive as unset (`2d2bdd04` / SERVE-MAXTOKENS-UNSET) and fills `max_model_len - seq_len` in `input_processor`. Keeping our clamp would fight that and conflict with #227/#228.

### Kept (this PR only)
- Full ROCm V1 sampler surface: `rocm_sample.hip` + allowed/token-mask helpers in `rocm_dense_basic.hip` + `RegisterOp` in `rocm_ops.hip`
- CMakeLists: add `rocm_sample.hip` only (no HIP `-O` / Triton comment thrash)
- CPU↔ROCm parity cases in `tests/vt/test_ops_sample.cpp` (skip when ROCm unlinked)
- USAGE note + FEATURES Gemma4 ROCm cell shortened under public-table limit

### Local gates (on this tip)
`check-device-leakage`, `check-public-doc-tables`, `check-doc-checkpoint`, `check-env-doc`, `check-agent-record`, `check-pr-size` — OK.

Please re-run CI on the force-pushed head. Happy to split further if review still wants sampler kernels alone vs docs.

**Comment by bakon11** [comment]
Rebased onto latest `main` (`f86e4cc0`) — tip `8ea62b21`. Local gates green (doc-checkpoint, env-doc, device-leakage, pr-size). Please re-run CI.

**Comment by bakon11** [comment]
## Tip-of-main rebuild (back on track)

Rebuilt as **single commit** on current `main` (`3286f6fd`) → tip `a418791e`.

- Payload only: `rocm_sample.hip` + ops registration + dense_basic hooks + CPU↔ROCm parity tests + minimal FEATURES/USAGE note
- **No** max_tokens clamp (already on main via `2d2bdd04`)
- Docs restored from main (no FEATURES/USAGE clobber)
- Local gates green: device-leakage, doc-checkpoint, public-doc-tables, env-doc, readme, commit-trailers, pr-size, agent-record

Please re-run CI. Prior red agent-record/pr-size were base-not-ancestor.

**Comment by bakon11** [comment]
CI hygiene 2026-08-11: rebased onto `origin/main` (60e71a0e). Local gates green (doc-checkpoint, env-doc, agent-record, pr-size, device-leakage). Land order remains #234 → #227 → #316 → #317.

**Comment by bakon11** [comment]
Rebased onto `origin/main` `c70f42b9` (main moved +2 during CI). Local gates green. Land order unchanged: #234 → #227 → #316 → #317. PR bodies refreshed (RandomSample note on #234; #228 closed in land path).

**Comment by bakon11** [comment]
@localai-org-maint-bot Status check against CHANGES_REQUESTED / localai-bot notes:

1. **CPU/ROCm parity tests** — `tests/vt/test_ops_sample.cpp` adds `HasRocm()` cases covering greedy/temperature/top-k-p/allowed_ids/logprobs and min_p/penalties (skip cleanly when no ROCm backend — inert on CPU CI).
2. **max_tokens→4096** — **removed** from this PR; main owns non-positive → unset (`2d2bdd04`). No `protocol.cpp` hunk on the tip.
3. PR is still **OPEN** (head `b067cd3a`) — treating the earlier “Merged.” comment as review prose, not GitHub merge state.

Sanitize ASan/TSan remain ambient-main class. Product jobs (build-test-cpu, cuda-fat, vulkan) were green on the last full run.
